### PR TITLE
[FLINK-15612][table] Refactor DataTypeLookup to DataTypeFactory

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

Refactors the `DataTypeLookup` for future use. It is integrated into `CatalogManager` with an adapted instance creation order for table environments.

## Brief change log

See commit messages.

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): yes
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/DataTypeFactory.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+	 */
+	default <T> DataType createDataType(Class<T> clazz) {
+		return DataTypeExtractor.extractFromType(this, clazz);
+	}
+
+	/**
+	 * Creates a RAW type for the given class. This type is a black box within the table ecosystem
+	 * and is only deserialized at the edges.
+	 *
+	 * <p>The factory will create {@link DataTypes#RAW(Class, TypeSerializer)} in cases where no serializer
```
Comment (dawidwys) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/DataTypeFactory.java]: This comment is a bit weird imo. Won't it always be a generic serializer? If not how is it different from `createDataType` (just from the interface perspective)?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/DataTypeFactory.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+	 *     // enrich the extraction with forcing using RAW types
+	 *     public @DataTypeHint(forceRawPattern = "scala.") Address address;
+	 *
+	 *     // enrich the extraction by specifying defaults
+	 *     public @DataTypeHint(defaultSecondPrecision = 3) Log log;
+	 *   }
+	 * }
+	 * </pre>
+	 */
+	default <T> DataType createDataType(Class<T> clazz) {
```
Comment (dawidwys) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/DataTypeFactory.java]: Do we need that default method? I don't think we will ever have a multiple implementations of that interface anyway, so one could implement everything here.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/DataTypeFactoryImpl.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+			ReadableConfig config,
+			@Nullable ExecutionConfig executionConfig) {
+		this.classLoader = classLoader;
+		this.executionConfig = createKryoExecutionConfig(classLoader, config, executionConfig);
+	}
+
+	@Override
+	public Optional<DataType> createDataType(String name) {
+		final LogicalType parsedType = LogicalTypeParser.parse(name, classLoader);
+		if (hasNestedRoot(parsedType, LogicalTypeRoot.UNRESOLVED)) {
```
Comment (dawidwys) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/DataTypeFactoryImpl.java]: I guess this is a very short term solution. We could think of having a method pretty much identical to `fromLogicalToDataType` that would accept a `DataTypeFactory` and fallback to `createDataType(UnresolvedIdentifier identifier)` for `UnresolvedUserDefinedType`.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/DataTypeFactoryImpl.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+			throw new TableException("User-defined types are not supported yet.");
+		}
+		return Optional.of(fromLogicalToDataType(parsedType));
+	}
+
+	@Override
+	public Optional<DataType> createDataType(UnresolvedIdentifier identifier) {
+		if (!identifier.getDatabaseName().isPresent()) {
+			return createDataType(identifier.getObjectName());
```
Comment (dawidwys) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/DataTypeFactoryImpl.java]: Have you thought how you want to pass `currentCatalog` & `currentDatabase` for the identifier resolution?

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/CatalogManager.java [hunk 5]
```diff
@@ -68,19 +72,88 @@
...
+		currentCatalogName = defaultCatalogName;
+		currentDatabaseName = defaultCatalog.getDefaultDatabase();
 
-		this.temporaryTables = new HashMap<>();
+		temporaryTables = new HashMap<>();
 		// right now the default catalog is always the built-in one
-		this.builtInCatalogName = defaultCatalogName;
+		builtInCatalogName = defaultCatalogName;
+
+		typeFactory = new DataTypeFactoryImpl(classLoader, config, executionConfig);
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/CatalogManager.java]: This is already prepared for looking up types in the future which requires access to the catalog manager itself. But we can also postpone this until we really need it.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/DataTypeFactoryImpl.java (hunk 4 above)
Comment (twalthr) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/DataTypeFactoryImpl.java]: The class will get access to catalog manager in the future

Comment (twalthr) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/DataTypeFactoryImpl.java]: The class will get access to catalog manager. It is basically only a component of catalog manager.

**Comment by twalthr** [comment]
Thanks for the review @dawidwys. I addressed your feedback. I also added one more commit that finializes this issue.

**Comment by twalthr** [comment]
Thank you @dawidwys. Merging...

**Pull request closed without merging**